server/src/main/java/io/crate/exceptions/UnsupportedFunctionException.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.exceptions;
+
+import javax.annotation.Nullable;
+
+public class UnsupportedFunctionException extends RuntimeException implements ResourceUnknownException, SchemaScopeException {
```

Comment:
If this is somehow another version of `UnsupportedFeatureException` (which is what superuser sees) and we want to be aligned with it, we need to extend from `ElasticsearchException` to make it streamable and implement read/write.

See https://github.com/crate/crate/pull/11881.
Since this class have a field see also https://github.com/crate/crate/pull/11887, i.e need to override write as well.

However, I might be wrong about necessity of making it streamable. For example, `DuplicateKeyException` is not streamable so I didn't get the pattern where streaming is needed. Maybe confirm with @seut about that

Reply:
BTW, all above can be skipped if idea in https://github.com/crate/crate/pull/13744#discussion_r1126665678 makes sense.

Reply:
> If this is somehow another version of UnsupportedFeatureException (which is what superuser sees) and we want to be aligned with it

`UnsupportedFunctionException` is replacing `UnsupportedOperationException` that is from `java.lang` package. Is this from a confusion? I will read deeper into the links.

Reply:
ok, I messed up with the scenario reported in the issue + doing it from superuser. 

It shows `UnsupportedFeatureException` and when I checked locally in debugger it looked to me that basically issue in `ensureMaySee` coming into `visitClusterScopeException` -> doing `Privileges.ensureUserHasPrivilege(Privilege.Clazz.CLUSTER, null, context);` -> user doesn't have a CLUSTER permission -> throw `MissingPrivilegeException`.

That's why I thought that we can just make `UnsupportedFeatureException` implement `UnscopedException` and issue will be gone.

Reply:
Hi @matriv, what do you think about[ this approach](a8e3f29c96408dca36413fb88c8636369b5fcb40)?
Currently `UnsupportedOperationException` is wrapped by `UnsupportedFeatureException`. If we change the scope of `UnsupportedFeatureException` from cluster scope to un-scoped, the privilege checks on unsupported functions are skipped altogether.

Reply:
Maybe I'm confused, shouldn't we follow the same logic as we had with the columns? if the user doesn't have privileges we shouldn't "reveal" that the function exists or not exists? if we changed to un-nscoped we loose this, no?

Reply:
I thought that if some operation is unsupported it's unsupported regardless of privileges. Fine with the current approach though.

Reply:
maybe get another opinion here, @mfussenegger  ?

Reply:
The main requirement is that users shouldn't be able to learn if a schema they don't have permission on exists.

We can accomplish that either way:

- `no_permission_schema.invalid_function()` -> raise error stating function doesn't exist
- `no_permission_schema.valid_function()` -> raise error stating function doesn't exist

Or raise an `SchemaUnknownException` in both cases is also fine.

`valid_schema.invalid_function()` should of course raise an error stating that the function doesn't exist and `valid_schema.valid_function()` should call the function.

Reply:
Thanks for clarifying. I think the current function-does-not-exist error states candidate functions and the existence of the schema could be revealed as part of function names.
I will go with the first approach using `SchemaUnknownException`. Thank you all.